### ISSUE TITLE
dFrameNumber in the firmware

### ISSUE THREAD

**Issue by 931436690** [Aharoni-Lab/Miniscope-DAQ-Cypress-firmware#5]
Hi,
I think dFrameNumber of DAQ should sending 3 bytes. 2 bytes of dFrameNumber, sending from DAQ to PC software, has max number, 65535. When camera using more  than 36 minutes, 2 bytes of dFrameNumber will lead to the negetave number of dropped frames, for example -65528.
I suggest to modify the firmware code like this:

                    else if (wValue == CY_FX_UVC_PU_CONTRAST_CONTROL) {
                    	glEp0Buffer[0] = CY_U3P_DWORD_GET_BYTE0(dFrameNumber);
			glEp0Buffer[1] = CY_U3P_DWORD_GET_BYTE1(dFrameNumber);
			glEp0Buffer[2] = CY_U3P_DWORD_GET_BYTE2(dFrameNumber);//adding one byte
                    	CyU3PUsbSendEP0Data (3, (uint8_t *)glEp0Buffer);//three bytes sending
                    }

**Comment by ximion** [comment]
Admittedly, I haven't looked very closely at this yet, but I don't think this solution would work since the contrast control value, according to the UVC specification, is a 2-byte unsigned integer and therefore has a maximum value of 65535 (unlike, for example, the exposure time control which is 4-byte unsigned).

![grafik](https://user-images.githubusercontent.com/417468/88899386-47aca400-d24e-11ea-8d39-2fef58565607.png)

Because it's unsigned though, I wonder how you actually got a negative number out here....

**Comment by 931436690** [comment]
Yes, it doesn't work, I just test it. We should change another way to send  a large number.
Next time, I will make sure the result first, then give the solution.

**Comment by daharoni** [comment]
This is one of many reasons why I am working on moving away from using UVC and instead working on implementing a more generic USB protocol. Will use the USB control channel for configuring the DAQ and Miniscope and a Bulk transfer endpoint (very similar to UVC) to transmit imaging data. Bulk transfer packets, and therefore, frames will be accurately timestamped.

I am also working on a methods for synchronizing multiple DAQs over USB so all incoming data will be accurately timestamped.

Hoping to have something running in the next week or two.

**Comment by 931436690** [comment]
I'm looking forward to hearing good news. Without the affect of UVC control, we can get stable timeStampes.

**Comment by ximion** [comment]
I've actually fixed this in https://github.com/Aharoni-Lab/Miniscope-DAQ-Cypress-firmware/commit/6c9596c86fb2f1a6494bda383a0e647df38e2674 - by using `UVC_CT_EXPOSURE_TIME_ABSOLUTE_CONTROL` you should get frame numbers which can go a lot higher than before :-)
